hypothesis-python/src/hypothesis/internal/conjecture/junkdrawer.py

```diff
@@ -199,48 +203,89 @@ class LazySequenceCopy:
         self.__len = len(values)
-        self.__mask = None
+        self.__mask: Optional[Dict[int, int]] = None
+        self.__popped_indices: Optional[SortedList] = None
 
     def __len__(self) -> int:
-        return self.__len
+        if self.__popped_indices is None:
+            return self.__len
+        return self.__len - len(self.__popped_indices)
 
-    def pop(self) -> int:
+    def pop(self, i: int = -1) -> int:
         if len(self) == 0:
             raise IndexError("Cannot pop from empty list")
-        result = self[-1]
-        self.__len -= 1
+        i = self.__underlying_index(i)
+
+        v = None
         if self.__mask is not None:
-            self.__mask.pop(self.__len, None)
-        return result
+            v = self.__mask.pop(i, None)
+        if v is None:
+            v = self.__values[i]
+
+        if self.__popped_indices is None:
+            self.__popped_indices = SortedList()
+        self.__popped_indices.add(i)
+        return v
 
     def __getitem__(self, i: int) -> int:
-        i = self.__check_index(i)
+        i = self.__underlying_index(i)
+
         default = self.__values[i]
         if self.__mask is None:
             return default
         else:
             return self.__mask.get(i, default)
 
     def __setitem__(self, i: int, v: int) -> None:
-        i = self.__check_index(i)
+        i = self.__underlying_index(i)
         if self.__mask is None:
             self.__mask = {}
         self.__mask[i] = v
 
-    def __check_index(self, i: int) -> int:
+    def __underlying_index(self, i: int) -> int:
         n = len(self)
         if i < -n or i >= n:
             raise IndexError(f"Index {i} out of range [0, {n})")
         if i < 0:
             i += n
         assert 0 <= i < n
+
+        if self.__popped_indices is not None:
+            i = self.__underlying_index_for_pops(i)
+
         return i
 
+    def __underlying_index_for_pops(self, i):
+        # given an index i in the popped representation of the list, compute
+        # its corresponding index in the underlying list. given
+        #   l = [1, 4, 2, 10, 188]
+        #   l.pop(3)
+        #   l.pop(1)
+        #   assert l == [1, 2, 188]
+        #
+        # we want l[i] == self.__values[f(i)], where f is this function.
+        assert len(self.__popped_indices) <= len(self.__values)
+
+        # we know what indices have been popped. count the total gap (number of
+        # "free indices") between these these, which is the elements remaining.
+        # stop whenever the gap is bigger than we need it to be and compute
+        # exactly where in the gap the expected index would be.
+        #
+        # This relies on popped_indices being sorted.
+        gap = 0
+        for n1, n2 in pairwise(
+            chain([-1], self.__popped_indices, [len(self.__values)])
+        ):
+            prev_gap = gap
+            gap += n2 - n1 - 1
+            if gap > i:
+                return n1 + (i - prev_gap) + 1
+
```

Comment:
I think we only need to account for the number of pops before the target index in the sorted list, and thus this can be simplified to something roughly like
```
        for idx in self.__popped_indices:
            if idx > i:
                break
            i += 1
        return i
```

This would also give the option of using bisect to find a close starting point before the linear search, which may be worthwhile if there are many pops. The "close starting point" will be the number of pops before the *input* index, which is always less than or equal to the number in the target index.

Reply:
ahh, thanks! I was sure there was a simpler way but didn't pick up on this.

Reply:
:+1: (I hope I'm right, wouldn't want to send you on a wild goose chase :grin:  )